### PR TITLE
Public api (re-)design

### LINKED ISSUE #20: Design the "public API"
Create a design for the public API.

The public API should not expose any implementation details (like Tree-Sitter types). It should provide a good and easy to use abstraction (e.g. no `std::shared_ptr` or `std::variant` those are all implementation details). But the API should still be customizable enough for most use cases (e.g. adding custom lua functions to the interpreter).

### PR DESCRIPTION
**Pull request by Lythenas**
Redesign the public api (a.k.a. the headers that will be installed with the library) to hide implementation details and only expose the actual public api we want to provide.

This is needed to remove the tree-sitter header from the install target. But this is also just in general sensible. E.g. the syntax tree is irrelevant for the user of the library.

Closes #20

### CODE REVIEW

**Comment by ThomasWitte** [comment]
minilua::Interpreter

- getting the changed source code to display it in the editor; currently I have to reimplement applying the SourceChanges to get the code and the highlights
- run should return something like `enum class ExecResult {NOERROR, ERR_PARSE, ERR_RUNTIME}` in addition to the SourceChanges and possibly the result value (a lua chunk is basically a function body and could easily have return values). Currently I return std::pair<ExecResult, lua::rt::eval_result_t> from my similar dostring method
- parse does not return whether it was successful
- Is the environment reset between calls to run? Do I have to create a new interpreter instance to reset it?
- Why would returning a Value from run make it necessary to expose the syntax tree? A function value can and does hide its body's code behind a pointer

minilua::Value

- Table::operator[] would be nice
- there should be a shortcut to `std::get<minilua::String>(Value("hallo").get()).value` like `Value("hallo").as_std_string()`
- How can the source location information of a value be accessed? I need a way to check whether location information is available for this value and force a new value on it. E.g. `if (v.has_source()) interpreter.force_value(v, 2.4);`
- What is the lifetime of Values and their source? I need NativeFunctions that give the argument value to a callback for the interactive marker. If it is moved, it forces its new position to the value. This can happen long after the NativeFunction returned and the interpreter is reset. In this case, I need to be able to manually apply the source changes.
- Value has no operators: `Value a, b; Value c = a+b;` should be possible and preserve source location information

minilua::Environment

- Currently, there is a feature in my editor that highlights the currently executed statement. Maybe, the debug.sethook function could replace the current solution. Is there a way to call lua functions or make the lua standard library functions available to C++? Environment.get('debug.sethook')? Environment.call?
- Adding functions to the Environment is only possible for string keys. Tables (including the global table) in lua map Value->Value, so it should also be possible to e.g. `Environment().add(Value(3.5), Value(true));`
- Environment::operator[] would be nice

**Comment by ThomasWitte** [comment]
> I planned to use exceptions for that. Bur you're right, the return value is missing. I think this is a easier to use in the interpreter because you can just throw an exception somewhere and don't have to check and propagate the error everywhere.
> 
> If it is easier for you we could catch them in `run` and add the error to `EvalResult`.

You should catch the exceptions internally. I think failing to run some programs is expected behaviour for the run method and should therefore not throw.

> I wanted to use an exception for this too. Maybe returning a bool would be easier (but you could easily miss the return value). Tree-Sitter will (almost) never fail to parse anyway. It will just insert "missing" nodes to recover from errors. We have to check for them manually.

Same as above. It is absolutely expected to fail to parse some input and should not throw IMO. It should return false if any missing nodes are in the tree. Maybe make it [[nodiscard]].

> I don't like that operator in C++ because it will insert values if they are missing. But I guess for Lua-Values it does not matter that much because we would just insert `nil`. It might fill the map with a lot of useless `nil` values if you check a bunch of them. I can add the operator if you want.

Yes, the behavior of operator[] on maps is strange. Maybe you could only add the const& variant, so that getting values from the table is possible but not adding them (using []). My point is, that I'd like to have some comfort using Values in my C++ code (operators for numbers, easy conversion to string or double etc.)

> Yes I agree. For `String` I can easily add an implicit conversion operator (`operator std::string() const`) or a method. But what is the behavior of converting a non `String` `Value` to a `std::string`?

I would throw a type error or even assert and terminate the program. I would make it obvious that this is risky and the type should be checked beforehand, so better no operator string.

> I need to think about this a bit and come up with an easy to use interface. Currently the `Value`s are just plain values.

Ok, then I wait for updates :-)

> Do you wan't the ability force new values from outside of `NativeFunction` too? Or just from inside of `NativeFunction`s?

It must be possible from outside, as I register some callbacks that force values in NativeFunctions.

> I can add the operators but what is the behavior of e.g. adding two tables or `nil`? Should that throw an exception for everything except numbers?

For tables it depends on the metatable (not yet implemented) and for nil or strings I would throw or assert, see above.

> I think we can add similar functionality. I'm not familiar with `debug.sethook`. I will add some way to call lua functions. It might also be useful to call lua functions from `NativeFunction`s.

sethook can register a callback function that is called every time the line (in the script) changed during execution.

> While it might be possible to do `_G[3.5] = true`. It does not make sense because you can't reference `3.5` except through the table. IMO this is pretty useless. And inside the interpreter this will be just a normal table. So from lua you have the normal behavior. But IMO this is not useful to expose from C++.

Ok, I agree, it doesn't make much sense :-)

> > Environment::operator[] would be nice
> 
> Maybe. Same problem as with table.

Ok, here it is probably not necessary. What is the conceptual difference between Environment and table? Does Environment contain a table or is it a table?

**Comment by Lythenas** [comment]
Ok I will make the following changes:

- [ ] change `parse` to return a `ParseResult` that contains whether it was successful (probably implicitly convertible to bool) and also the positions/ranges with an error
- [ ] add a return value and a optional error to `EvalReult`
- [ ] rename `run` to `evaluate` (I think that is a better name, also in combination with `EvalResult`)
- [ ] add `force_value` to `Interpreter` and to `CallContext` (which is the parameter of `NativeFunction`). I probably also need to change/rethink `CallContext` a bit.

The rest might need some more discussion.

Maybe first a comment on what I based my design decisions on: I didn't really mirror the lua semantic/behavior (e.g. environment is just another table). I wanted to provide an interface that is easy to use in C++. That's why I started with an `Environment` that is basically a `std::map<std::string, Value>`. This was never intended to be used during the actual execution of the interpreter. Only to "configure" the interpreter.

For `Value& Table::operator[]` I think it makes the most sense to just follow what C++ does in its std lib (add a new value if it does not exist) even if I don't like it. But I will also provide `get` and `set` methods that can be used if you want more control.

I am fine with adding helper functions/operators to the specific value types (`minilua::String`, `minilua::Number`, etc). But I'm not sure if it really makes sense to also add them to the wrapping `minilua::Value` type. While it might be more convenient, we would basically be implementing (a small part of) the lua interpreter in C++ syntax.

However if we don't provide e.g. `Value operator+(const Value&, const Value&)` that means you would need to special case adding numbers and adding (meta)tables when you implement a `NativeFunction`. So it probably makes sense to add them to `Value` too. (The C-API for the official lua interpreter is IMO completely ridiculous). I will however throw exceptions everywhere because here it is the expected case that the types are compatible. Also I think `to_number` (etc.) functions make sense.

> > Do you wan't the ability force new values from outside of NativeFunction too? Or just from inside of NativeFunctions?
>
> It must be possible from outside, as I register some callbacks that force values in NativeFunctions.

So basically you want a function like `Interpreter::force_value(Value source_value, Value new_value)`.

Does/can this happen while the interpreter is still running?

**Comment by ThomasWitte** [comment]
> Maybe first a comment on what I based my design decisions on: I didn't really mirror the lua semantic/behavior (e.g. environment is just another table). I wanted to provide an interface that is easy to use in C++. That's why I started with an `Environment` that is basically a `std::map<std::string, Value>`. This was never intended to be used during the actual execution of the interpreter. Only to "configure" the interpreter.

Ok, makes sense, The existing lua::rt::Environment confused me and I thought it is used at runtime.

> For `Value& Table::operator[]` I think it makes the most sense to just follow what C++ does in its std lib (add a new value if it does not exist) even if I don't like it. But I will also provide `get` and `set` methods that can be used if you want more control.

You could also return a proxy object like vector<bool> does. Then it would be possible to clean up unused keys in its destructor (or lazyly add keys only if actually needed)

> I am fine with adding helper functions/operators to the specific value types (`minilua::String`, `minilua::Number`, etc). But I'm not sure if it really makes sense to also add them to the wrapping `minilua::Value` type. While it might be more convenient, we would basically be implementing (a small part of) the lua interpreter in C++ syntax.
> 
> However if we don't provide e.g. `Value operator+(const Value&, const Value&)` that means you would need to special case adding numbers and adding (meta)tables when you implement a `NativeFunction`. So it probably makes sense to add them to `Value` too. (The C-API for the official lua interpreter is IMO completely ridiculous). I will however throw exceptions everywhere because here it is the expected case that the types are compatible. Also I think `to_number` (etc.) functions make sense.

One more thing to point out is that you want the source location bookkeeping happen automatically. That means you need the operators. It doesn't really matter whether it is in Number or Value. You're right that generic code profits from having them in value, while having them in Number is probably safer (but then the exception is thrown while carelessly casting Value to Number, so it doesn't really matter, I think). Yes to exceptions.

> So basically you want a function like `Interpreter::force_value(Value source_value, Value new_value)`.
> 
> Does/can this happen while the interpreter is still running?

Currently, it does not happen, while the interpreter is still running. The callback event handlers and the interpreter run in the same thread, so thread safety should not be a problem. The semantics would also be unclear: All source changes (including these) are applied after execution finishes, otherwise you would have to rewind part of the execution?

File: include/MiniLua/MiniLua.hpp [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+ *
+ * If the nested `struct Impl` is not default constructible you have to provide
+ * the implementation of the copy-/move-constructors and -operators manually and
+ * can't use `= default`.
+ *
+ * `owning_ptr<T>` was chosen for the pointer type because it behaves like a
+ * normal `T` (but lives on the heap). It's move, copy and lifetime semantics
+ * are identical to the one of `T`.
+ *
+ * Is is also possible to choose another pointer type like `std::unique_ptr` or
```
Comment (tn4799) [inline include/MiniLua/MiniLua.hpp:75]: Change to "It is ..."

File: include/MiniLua/source_change.hpp [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+ * correct results you should only compare locations that were generated from
+ * the same source code.
+ */
+struct Location {
+    uint32_t line;
+    uint32_t column;
+    uint32_t byte;
+};
+
+constexpr auto operator==(Location lhs, Location rhs) noexcept -> bool {
```
Comment (tn4799) [inline include/MiniLua/source_change.hpp:27]: Maybe add a Testcase

File: include/MiniLua/source_change.hpp [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+
+/**
+ * Represents a range/span in source code.
+ */
+struct Range {
+    Location start;
+    Location end;
+};
+
+constexpr auto operator==(Range lhs, Range rhs) noexcept -> bool {
```
Comment (tn4799) [inline include/MiniLua/source_change.hpp:53]: maybe add testcase in the future

**Review by tn4799 — APPROVED** [review]
Some little improvements could be made in the future, especially with tests, but for now in my opinion its fine right now.

File: include/MiniLua/MiniLua.hpp (hunk 1 above)
Comment (Lythenas) [reply include/MiniLua/MiniLua.hpp:75]:
```suggestion
 * It is also possible to choose another pointer type like `std::unique_ptr` or
```

**Review by tn4799 — APPROVED** [review]
